lib/crypto.js

```diff
@@ -95,6 +96,25 @@ Hmac.prototype.digest = Hash.prototype.digest;
 Hmac.prototype._flush = Hash.prototype._flush;
 Hmac.prototype._transform = Hash.prototype._transform;
 
+// This implements Brad Hill's Double HMAC pattern from
+// https://www.nccgroup.trust/us/about-us/
+//   newsroom-and-events/blog/2011/february/double-hmac-verification/.
+// In short, it's near-impossible to write a reliable constant-time compare in a
+// high level language like JS, because of the many layers that can optimize
+// away attempts at being constant time.
+//
+// Double HMAC avoids that problem by blinding the timing channel instead. After
+// running the inputs through a second round of HMAC, we are free to
+// short-circuit comparison, because the time it takes to reach the
+// short-circuit has no relation to the similarity between the guessed digest
+// and the correct one.
+//
+// Note: hmac object can not be used after validate() method has been called.
+Hmac.prototype.validate = function(inputBuffer) {
+  var ah = new Hmac('sha256', this.key).update(this.digest()).digest();
```

Comment:
HMAC-SHA256 is a strong pseudorandom function, so you don't leak anything by using it in this fashion.  Re-using the same key for every comparison weakens the HMAC function in a very theoretical way, such that a forgery can be created at just the collision-resistance of the underlying hash. (birthday bound for a perfect hash)  For HMAC-SHA-256, guessing over a network, this is still practically impossible (2^128 guesses), and even then such a forgery is not a chosen-plaintext forgery.  You can avoid this by picking a new blinding key for each comparison if you really care, but it's not worth paying much cost for.